tty.c

```diff
@@ -109,101 +116,232 @@ void tty_setup_socketpair(int *pParentSock, int *pChildSock) {
+		flags |= O_NONBLOCK;
+	} else {
+		flags &= ~O_NONBLOCK;
+	}
+	if (fcntl(fd, F_SETFL, flags) == -1) {
+		err(1, "fcntl %d F_SETFL O_NONBLOCK", fd);
+	}
+}
+
 void tty_parent_cleanup() {
 	if (info.termfd >= 0) {
+		/* Drain any remaining data in the terminal buffer */
+		set_nonblock(STDOUT_FILENO, 0);
+		set_nonblock(info.termfd, 0);
+		struct buffer drain = {
+			.size = 0,
+		};
+
+		if (io_copy(STDOUT_FILENO, info.termfd, &drain) == -1 && errno != EIO) {
+			warn("copy tty -> stdout");
+		}
+
 		close(info.termfd);
+		info.termfd = -1;
 	}
 	if (info.stdinIsatty) {
 		tcsetattr(STDIN_FILENO, TCSADRAIN, &info.orig);
+		info.stdinIsatty = false;
 	}
 }
 
 void tty_set_winsize() {
 	struct winsize wsize;
-	if (ioctl(STDIN_FILENO, TIOCGWINSZ, (char*) &wsize) < 0) {
-		err(1, "reading window size");
-	}
-	if (ioctl(info.termfd, TIOCSWINSZ, (char*) &wsize) < 0) {
-		err(1, "writing window size");
+	if (info.stdinIsatty) {
+		if (ioctl(STDIN_FILENO, TIOCGWINSZ, (char*) &wsize) < 0) {
+			err(1, "reading window size");
+		}
+		if (ioctl(info.termfd, TIOCSWINSZ, (char*) &wsize) < 0) {
+			err(1, "writing window size");
+		}
 	}
 }
 
-bool tty_handle_sig(siginfo_t *siginfo) {
-	switch (siginfo->si_signo) {
-	case SIGWINCH:
-		if (!info.stdinIsatty) return false;
-		tty_set_winsize();
-		return true;
-	}
-	return false;
+static int tty_handle_sig(int epollfd, const struct epoll_event *ev, int fd, pid_t pid)
+{
+	siginfo_t siginfo;
+	sig_read(fd, &siginfo);
+
+	assert(siginfo.si_signo == SIGWINCH && "tty_handle_sig can only handle SIGWINCH");
+	tty_set_winsize();
+	return EPOLL_HANDLER_CONTINUE;
 }
 
-bool tty_parent_select(pid_t pid) {
-	const size_t buflen = 1024;
-	char buf[buflen];
-	fd_set readFds = info.rfds, writeFds = info.wfds;
-	bool rtn = false;
+static struct epoll_handler inbound_handler, outbound_handler, term_handler;
 
-	int rc = select(info.nfds, &readFds, NULL, NULL, NULL);
-	if (rc == 0) {
-		return false;
-	}
-	if (rc < 0) {
-		if (errno == EINTR) {
-			return false;
+static struct buffer inbound_buffer, outbound_buffer;
+
+static int tty_handle_io(int epollfd, const struct epoll_event *ev, int fd, pid_t pid)
+{
+	struct epoll_handler *handler = ev->data.ptr;
+
+	if (fd == inbound_handler.fd) {
+		if (ev->events & EPOLLIN) {
+			handler->ready |= READ_READY;
 		}
-		err(1, "select");
-	}
-	struct timeval immediate = {0};
-	if (select(info.nfds, NULL, &writeFds, NULL, &immediate) < 0) {
-		return false;
-	}
-	if (FD_ISSET(STDIN_FILENO, &readFds) && FD_ISSET((unsigned long) info.termfd, &writeFds)) {
-		ssize_t nread = read(STDIN_FILENO, buf, buflen);
-		if (nread > 0) {
-			if (write(info.termfd, buf, (size_t) nread) < 0) {
-				warn("writing to terminal");
-			}
-		} else {
-			if (nread < 0) {
-				warn("reading from stdin");
-			}
-			FD_CLR(STDIN_FILENO, &info.rfds);
-			if (write(info.termfd, &(char){4}, 1) < 0) {
-				warn("writing EOT to terminal");
+		if (ev->events & EPOLLHUP) {
+			send_veof(handler->peer_fd);
```

Comment:
Where does send_veof come from?  Trying to look it up?

Reply:
this was caused by a wrong conflict resolution with rebase -- see updated diff